source/dev/coding_guide.rst

```diff
@@ -0,0 +1,441 @@
+that cannot be replaced, overridden, or abstracted away in some other way.
+
+Impurity is toxic: if method ``foo()`` depends on non-deterministic or side-effecting
+method ``bar()``, then ``foo()`` becomes non-deterministic or side-effecting as well.
+Eventually, we may end up poisoning the entire codebase. Multiply all these problems by
+the size of a complex real-life application, and we’ll find ourselves encumbered with a
+hard to maintain codebase full of smells, anti-patterns, secret dependencies, and all
+sorts of ugly and unpleasant things.
+
+However, impurity is inevitable; any real-life application must, at some point, read and
+manipulate state by interacting with the environment, databases, configuration files,
+web services, or other external systems. So instead of aiming to eliminate impurity
+altogether, it’s a good idea to limit these factors, avoid letting them poison your
+codebase, and break hard-coded dependencies as much as possible, in order to be able to
+analyze and unit test things independently.
+
+
+Problem: Global Mutable State
++++++++++++++++++++++++++++++
+
+Static Properties and Fields
+****************************
+Mutable static properties and fields are global state! They can hide the information
+required for a method to get its job done, introduce non-determinism, or promote
+extensive usage of side effects. Functions that read or modify mutable global state are
+inherently impure.
+
+Example:
+
+.. code-block:: python
+
+    if (SmartHomeSettings.cost_saving_enabled is False):
+        _SwimmingPoolController.heat_water()
+
+What if the ``heat_water()`` method doesn’t get called when we are sure it should have
+been? Since any part of the application might have changed the ``cost_saving_enabled``
+value, we must find and analyze all the places modifying that value in order to find out
+what’s wrong. Also, as we’ve already seen, it is not possible to set some static
+properties for testing purposes; they are read-only, but still non-deterministic.
+
+Note that global *constants* that never change do not cause non-determinism or
+side-effects.
+
+.. code-block:: python
+
+    def circumference(radius):
+        return 2 * math.pi * radius # Because math.pi is a global constant, this is still a pure function!
+
+
+Singletons
+**********
+Essentially, the Singleton pattern is just another form of the global state. Singletons
+promote obscure APIs that lie about real dependencies and introduce unnecessarily tight
+coupling between components. They also violate the Single Responsibility Principle
+because, in addition to their primary duties, they control their own initialization and
+lifecycle.
+
+Singletons can easily make unit tests order-dependent because they carry state around
+for the lifetime of the whole application or unit test suite. Have a look at the
+following example:
+
+.. code-block:: python
+
+    def get_user(user_id):
```

Comment:
I think the original version of this is almost right, except for the casing, and would change the `contains_key` to be `user_id in user_cache.instance`

Reply:
done